portal-api/src/v1/users/sign-in-link.service.js

```diff
@@ -17,37 +17,47 @@ class SignInLinkService {
   async createAndEmailSignInLink(user) {
     const { _id: userId, email: userEmail, firstname: userFirstName, surname: userLastName } = user;
 
-    const signInCode = this.#createSignInCode();
-    await this.#saveSignInCodeHashAndSalt({ userId, signInCode });
+    const signInToken = this.#createSignInToken();
+    await this.#saveSignInTokenHashAndSalt({ userId, signInToken });
 
     return this.#sendSignInLinkEmail({
-      signInLink: `${PORTAL_UI_URL}/login/sign-in-link?t=${signInCode}`,
+      signInLink: `${PORTAL_UI_URL}/login/sign-in-link?t=${signInToken}`,
       userEmail,
       userFirstName,
       userLastName,
     });
   }
 
-  #createSignInCode() {
+  async isValidSignInToken({ userId, signInToken }) {
+    const user = await this.#userRepository.findById(userId);
+    if (!user.hash || !user.salt) {
+      throw new Error('User does not have a valid sign in token.');
+    }
+
+    const { hash, salt } = user.signInToken;
+    return this.#hasher.verifyHash({ target: signInToken, hash: Buffer.from(hash, 'hex'), salt: Buffer.from(salt, 'hex') });
```

Comment:
Maybe something you want to worry about in the next PR, or later, but there's a bit of a mismatch between saving and reading the hash and salt

When we save the hash and salt, we give the repository the bytes and it decides to convert them to hex strings to be stored in the db.
When we retrieve the hash and salt, the repository gives us the hex strings and we need to know that they need to be converted back to bytes

Ideally it would be consistent, i.e.
1. we give the repository bytes to save and it gives us back bytes when we read, or
2. we give the repository hex strings to save and it gives us back hex strings when we read

Personally I prefer (1) because the bytes are the actual hash and salt and the hex strings are just the format we choose to persist that info in the db

Reply:
Updated to match first option